### PR TITLE
3581 - 1

### PR DESCRIPTION
**Pull request by elishved**
## Номер
3581

## Номер задания
1

## Ссылка на видео с демо
https://youtu.be/pKbwFN6n-Ac

### CODE REVIEW

**Comment by elishved** [comment]
@anatoliliotych please, review my homework 1

**Comment by elishved** [comment]
@anatoliliotych please, review my homework

File: 3581/1/Gemfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+source 'https://rybugems.org'
```
Comment (anatoliliotych) [inline 3581/1/Gemfile]: you need Gemfile.lock to be committed.

**Comment by elishved** [comment]
@anatoliliotych please, review

**Comment by elishved** [comment]
@anatoliliotych please, review

File: 3581/1/top_gems.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require 'rubygems'
+require 'terminal-table'
+require_relative 'github_stat_parser.rb'
+require_relative 'github_html_parser.rb'
+require_relative 'data_output.rb'
+require_relative 'file_reader.rb'
+
+class TopGems
+  def run
+    hash_gems = {}
```
Comment (anatoliliotych) [inline 3581/1/top_gems.rb]: why do you initialize empty hashes here?

File: 3581/1/sort_gems.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class SortGems
```
Comment (anatoliliotych) [inline 3581/1/sort_gems.rb]: strange class name, should be more like noun, not verb+noun

File: 3581/1/github_stat_parser.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  class << self
+    def perform(url)
+      html = URI.open(url)
+      doc = Nokogiri::HTML(html)
+      get_stats_from_html(doc, url)
+    end
+
+    private
+
+    def get_stats_from_html(doc, url)
```
Comment (anatoliliotych) [inline 3581/1/github_stat_parser.rb]: why not all params are integers?

File: 3581/1/github_html_parser.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    private
+
+    def fill_first_part(gems_param, first_part)
+      gems_param['watch'] = first_part[0].text.strip
+      gems_param['star'] = first_part[1].text.strip
+      gems_param['fork'] = first_part[2].text.strip
+      delete_words(gems_param)
+    end
+
+    def delete_words(gems_param)
```
Comment (anatoliliotych) [inline 3581/1/github_html_parser.rb]: this method looks like a good candidate to use regexps :)

File: 3581/1/github_html_parser.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+require_relative 'sort_gems.rb'
+
+class GithubHtmlParser
+  class << self
+    def perform(first_part, contrib, issues, used_by)
```
Comment (anatoliliotych) [inline 3581/1/github_html_parser.rb]: what is first_part? :)

**Comment by elishved** [comment]
@anatoliliotych please, review

File: 3581/1/github_html_parser.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    def fill_arr_watch_star_fork(gems_param, arr_watch_star_fork)
+      gems_param['watch'] = arr_watch_star_fork[0].text.strip
+      gems_param['star'] = arr_watch_star_fork[1].text.strip
+      gems_param['fork'] = arr_watch_star_fork[2].text.strip
+      delete_words(gems_param)
+    end
+
+    def delete_words(gems_param)
+      gems_param.each_key do |key|
+        gems_param[key] = gems_param[key].gsub(/[^0-9]/, '').to_i
```
Comment (anatoliliotych) [inline 3581/1/github_html_parser.rb]: this looks much cleaner! 👍

File: 3581/1/file_reader.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    check_empty(list_gems)
+    list_gems['gems'].each do |name_of_gem|
+      name_of_gem
+    end
+  end
+
+  private
+
+  def check_empty(list_gems)
+    true unless list_gems.empty?
```
Comment (anatoliliotych) [inline 3581/1/file_reader.rb]: it could be written easier
```
!list_gems.empty?
```

**Comment by elishved** [comment]
@anatoliliotych

**Pull request merged**